gobb_inspect.1.md
# NAME

gobb_inspect - inspect positions of Gobblet Gobblers

# SYNOPSIS

gobb_inspect [OPTION]... [POSITION]

# DESCRIPTION

`gobb_inspect` shows the analysis data generated by `gobb_analyze` as text messages.

When `gobb_analyze` is launched, it first searches the current directory for a data file
`gobb_analyzer_<GENERATION>.dat`.
If found, it loads a file with the largest generation number.
`gobb_inspect` reads the entire file with 3GB, so that it may take for a while.

After the loading of the data file, `gobb_inspect` prints _the current position_, possible moves of
the current position and a prompt `'gobb_inspect'>` to standard out.
It is recommended to use a terminal with a minimum width of 100, and a height of 48.
If the argument `POSITION` is given, the current position is set to the specified value.
Otherwise, it is set to 0, which is the initial position of the game.

The program interactively accepts input from standard in and prints the result messages to standard out.
It continues to perform the above actions until the program is terminated.
The command consumes about 3GB of memory at run time.

# INTERACTIVE COMMANDS

## Position

sp, show-position
: show the current position.

gp ID, goto-position ID
: go to the position ID.

## Move

sm, show-moves
: show possible moves.

smb, show-movebacks
: show possible retrograde moves.

m NUM, move NUM
: execute the movement of the possible move NUM.

mb NUM, moveback NUM
: execute the movement of the possible retrograde move NUM.

## History

`gobb_inspect` manages a simple history table of visited positions.
The maximum size of the history table is 32.

sh, show-history
: show the history table.

gh NUM, goto-history NUM
: go to the position of the history NUM.

n, next
: go to the next position of the history.

p, previous:
: go to the previous position of the history.

## Miscellaneous

?, help
: print help messages.

exit
: exit the program.

Sending EOF (`Ctrl-D` or `Ctrl-Z`) also exits the program.

# NAMING THE SQUARES

The ordinal directions is used for identifying each square.

* NW (northwest).
* N (north).
* NE (northeast).
* W (west).
* Center (center).
* E (east).
* SW (southwest).
* S (south).
* SE (southeast).
* Out (out of the board)

# ANALYSIS STATUS

Each position is marked with one of the following status codes.

* Unfixed
  - the game never ends as long as both player play perfectly.
* Lost
  - the player having the current turn lost the game, or
  - the player having the current turn will lose the game, even if he plays perfectly.
* LostStalemate
  - the player having the current turn must pick up his piece on the board, but it causes that three pieces
    of the opponent are lined up in a row (the position ID 822009771 for example).
* Won
  - the player having the current turn will win the game, as long as he plays perfectly.
* WonStalemate
  - the player having the current turn will win the game, and the opponent will lose as a stalemate.
* Contradictory
  - at the beginning of the turn, three pieces of the player having the current turn have already been lined up
    in a row (the position ID 707034748 for example), or
  - the player having the current turn has not placed any piece on the board yet, but the opponent has placed
    two or more pieces (the position ID 1 for example), or
  - at the beginning of the turn, the player having the current turn has placed one or more pieces on the board,
    but the opponent has not placed any piece on the board yet (the position ID 379 for example).

# OPTIONS

`gobb_inspect` recognizes the command line options listed below.

-c
: Print colored pieces using escape sequences.
: This is the default in case that standard out refers to a terminal.

-C
: Print colorless pieces.
: This is the default in case that standard out doesn't refer to a terminal.

-d DIR
: Read the analysis data file at DIR instead of the current directory.

-g GENERATION
: Specify a generation number of the data file to be loaded.
: If also `-d` option is given, `gobb_inspect` loads the file at the specified directory.
: Otherwise it loads the file at the current directory.

--help
: Show help messages, then exit.

--version
: Show the version, then exit.

# NOTES

`gobb_inspect` displays colored pieces using escape sequences.
On Windows, Command Prompt and PowerShell support escape sequences, but they are disabled by default.
If you would like to make Command Prompt and PowerShell recognize escape sequences, set the registory

    HKEY_CURRENT_USER\Console\VirtualTerminalLevel

to 1 (REG_DWORD) and then execute a terminal program.

# SEE ALSO

`gobb_analyze(1)`


gobb_inspect.cpp
//
// Copyright (C) 2022 Motoyuki Kasahara.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>
//

#include <cstring>
#include <exception>
#include <iostream>
#include <string>
#include "analysis_data_file_handler.hpp"
#include "gobb_inspect_processor.hpp"
#include "position_text_creator.hpp"
#include "string_to_uint.hpp"
#include "inspector.hpp"
#include "version.hpp"

#if defined(_WIN32)
#include <io.h>
#elif defined(HAVE_UNISTD_H)
extern "C" {
#include <unistd.h>
}
#endif

//
// Print the help message.
//
void print_help_message() {
    std::cout << "Usage: gobb_inspect [OPTION...] [POSITION-ID]" << std::endl;
    std::cout << "Options:" << std::endl;
    std::cout << "  -c          print pieces in color on the terminal" << std::endl;
    std::cout << "  -C          do not print pieces in color on the terminal" << std::endl;
    std::cout << "  -d DIR      load an analysis data file in DIR (default: .)" << std::endl;
    std::cout << "  -g NUM      load analysis data file of the NUM'th generation" << std::endl;
    std::cout << "              (default: the latest generation stored)" << std::endl;
    std::cout << "  --help      print this help, then exit" << std::endl;
    std::cout << "  --version   print version information, then exit" << std::endl;
}

//
// Print the version information.
//
void print_version() {
    std::cout << "Gobb Analyzer version " << GOBB_ANALYZER_VERSION << std::endl;
}

//
// Print "try 'gobb_analyze --help' ..." message.
//
void print_hint(const char* argv0) {
    std::cout << "Try '" << argv0 << " --help' for more information." << std::endl;
}

//
// Main.
//
int main(int argc, char* argv[]) {
    //
    // Parses command line arguments.
    //
    std::string dataDir;
    unsigned long generation = 0u;
    bool opt_c = false;
    bool opt_d = false;
    bool opt_g = false;
#if defined(_WIN32)
    opt_c = _isatty(1);
#elif defined(HAVE_UNISTD_H)
    opt_c = isatty(1);
#endif

    int optind = 1;
    while (optind < argc) {
        if (argv[optind][0] != '-' || argv[optind][1] == '\0') {
            break;
        }

        char ch = argv[optind][1];
        if (ch == '-' && argv[optind][2] == '\0') {
            optind++;
            break;
        } else if (ch == 'c') {
            opt_c = true;
            optind++;
        } else if (ch == 'C') {
            opt_c = false;
            optind++;
        } else if (ch == 'd') {
            opt_d = true;
            const char* optarg;
            if (argv[optind][2] == '\0') {
                if (optind + 1 >= argc) {
                    std::cerr << argv[0] << ": missing argument to option '-d'" << std::endl;
                    print_hint(argv[0]);
                    return 1;
                }
                optarg = argv[optind + 1];
                optind += 2;
            } else {
                optarg = argv[optind] + 2;
                optind++;
            }
            dataDir = std::string(optarg);
        } else if (ch == 'g') {
            const char* optarg;
            opt_g = true;
            if (argv[optind][2] == '\0') {
                if (optind + 1 >= argc) {
                    std::cerr << argv[0] << ": missing argument to option '-g'" << std::endl;
                    print_hint(argv[0]);
                    return 1;
                }
                optarg = argv[optind + 1];
                optind += 2;
            } else {
                optarg = argv[optind] + 2;
                optind++;
            }
            if (!string_to_uint(optarg, generation) || generation > MaxGeneration) {
                std::cerr << argv[0] << ": invalid generation '" << optarg << "'" << std::endl;
                print_hint(argv[0]);
                return 1;
            }
        } else if (std::strcmp(argv[optind], "--help") == 0) {
            print_help_message();
            return 0;
        } else if (std::strcmp(argv[optind], "--version") == 0) {
            print_version();
            return 0;
        } else {
            std::cerr << argv[0] << ": invalid option '-" << ch << "'" << std::endl;
            print_hint(argv[0]);
            return 1;
        }
    }

    if (optind + 1 < argc) {
        std::cerr << argv[0] << ": too many arguments" << std::endl;
        print_hint(argv[0]);
        return 1;
    }

    PositionId posId;
    if (optind + 1 == argc) {
        if (!string_to_uint(argv[optind], posId)) {
            std::cerr << argv[0] << ": invalid position '"  << argv[optind] << "'" << std::endl;
            print_hint(argv[0]);
            return 1;
        }
        optind++;
    } else {
        posId = InitialPositionId;
    }

    //
    // Creates a GobbInspectProcessor instance and starts interactive processing.
    //
    PositionAsciiCreator textCreator(opt_c);

    try {
        AnalysisDataFileHandler fileHandler;
        if (opt_d) {
            fileHandler = AnalysisDataFileHandler(dataDir);
        }

        Inspector inspector;
        if (opt_g) {
            if (!inspector.load(fileHandler, static_cast<Generation>(generation))) {
                std::cerr << "failed to load the analysis data file of the specified generation" << std::endl;
                return 1;
            }
        } else {
            if (inspector.load_latest(fileHandler) == InvalidGeneration) {
                std::cerr << "failed to load an analysis data file" << std::endl;
                return 1;
            }
        }

        GobbInspectProcessor processor(inspector, textCreator, posId);
        processor.do_main_loop();
    } catch (std::exception& err) {
        std::cerr << "an exception raised, " << err.what() <<  std::endl;
        return 1;
    }

    return 0;
}
